### DISCUSSION TITLE
How to link to pyodide-maintained openblas/clapack when building package out-of-tree?

### DISCUSSION THREAD

**Discussion by hzhangxyz** [pyodide/pyodide#4340] [Q&A] [answered]
Hello, I am following [this](https://pyodide.org/en/stable/development/building-and-testing-packages.html) to [build](https://github.com/USTC-TNS/TAT/actions/runs/7205486200) my package out of tree and it works well currently. However I am linking my package to a emscripten clapack built by myself, while pyodide has its prebuilt clapack(pyodide 0.23.x) or openblas(pyodide 0.24.x) already, so I want to know how to modify my setup.py or pyproject.toml or something else to link to pyodide-maintained openblas/clapack directly?

There is a related discussion https://github.com/pyodide/pyodide/discussions/2201, but it is about building in-tree, making use of meta.yaml, which is not the same to out-of-tree building.

**Comment by ryanking13** [comment, marked as answer]
I have some plans to release the library build artifacts (header files and libraries) that Pyodide uses so that people who build packages out-of-tree can use it. But it is not done yet.

But especially for clapack and openblas, you can check the [xbuildenv archive](https://github.com/pyodide/pyodide/releases/tag/0.25.0a2). If you navigate to the `pyodide-root/packages/.libs/` path in the archive, you'll find the header files and shared libraries from building the blas library, which you can link to when building your packages.

**Reply by hzhangxyz** [reply]
Ohhh, thanks for your answer.

Currently, I am compiling clapack to three `.a`: `liblapack.a`, `libblas.a` and `libf2c.a`, and then link them with my own `.o` alltogether to get a python extension.

Based on your advice, download xbuildenv archive, and then I try:

1. link `libopenblas.a` directly, just like link my self-built `.a`, everything works well.

2. link `libopenblas.so` directly, compile correctly. But I must run `await pyodide.loadPackage("openblas");` manually before loading my package. Or I need to add it to the dependence of my package. So, how to detect whether I am building package for pyodide in `setup.py`?

**Reply by ryanking13** [reply]
> link libopenblas.so directly, compile correctly. But I must run await pyodide.loadPackage("openblas"); manually before loading my package. Or I need to add it to the dependence of my package. So, how to detect whether I am building package for pyodide in setup.py?

There are two options (for now) to make Pyodide aware of the shared library dependencies. 

1. Copy shared libraries into the wheel file.

```bash
$ pip install auditwheel-emscripten
$ pyodide auditwheel copy <your_wheel_file> --libdir <directory where openblas.so located> ...
```

This command copies the openblas.so into the wheel and Pyodide is aware that it should load the library when you load your wheel.

2. Add your wheel into pyodide-lock.json.

```
$ pip install pyodide-lock
$ pyodide lockfile add-wheels <your wheel file> ...
```

This command adds your wheel (and its dependencies) to the existing pyodide-lock.json. (cc @joemarshall)

**Reply by hzhangxyz** [reply]
Thanks for you answer, auditwheel works great.

I want to know if my wheel file can omit embedding dynamic libraries, just like the way pyodide maintains scipy, which doesn't embed openblas but as a dependency. When installing scipy, openblas is downloaded and installed first

**Reply by ryanking13** [reply]
> I want to know if my wheel file can omit embedding dynamic libraries, just like the way pyodide maintains scipy, which doesn't embed openblas but as a dependency. When installing scipy, openblas is downloaded and installed first

Yeah, that's what `pyodide-lock.json` file does. It specifies openblas as a scipy's dependency, so Pyodide can load openblas before loading scipy.

Unfortunately, `pyodide lockfile add-wheels` does not handle shared library dependencies, at least for now... So for now, building packages with an in-tree build is the most straightforward way to explicit specify shared library dependencies.
